Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+                                ypk, epk_sq = focus_data_in_detector_mask(signal, error, peak_mask, non_bg_mask, ixpk)
+                                ixlo_opt, ixhi_opt = find_peak_limits(ypk, epk_sq, min(ixlo, ixlo_opt),
+                                                                      max(ixhi, ixhi_opt))
+                    if update_peak_pos:
+                        # find det
+                        idet = np.argmax(signal[:, :, ixlo_opt:ixhi_opt].sum(axis=2)[peak_mask])
+                        det = dets[peak_mask][idet]
+                        irow, icol = np.where(dets == det)
+                        # update tof
+                        imax = np.argmax(ypk[ixlo_opt:ixhi_opt]) + ixlo_opt
+                        # replace last added peak
+                        irows_delete.append(pk_ws_int.getNumberPeaks() - 1)
+                        self.child_AddPeak(PeaksWorkspace=pk_ws_int, RunWorkspace=ws, TOF=xpk[imax],
+                                           DetectorID=int(det))
+                        pk = pk_ws_int.getPeak(pk_ws_int.getNumberPeaks() - 1)
+                    # normalise by bin width before final integration
+                    dx = np.diff(xpk)
+                    ypk[1:] = ypk[1:] * dx
+                    ypk[0] = ypk[0] * dx[0]  # assume first has same dx as adjacent bin
+                    epk_sq[1:] = epk_sq[1:] * (dx ** 2)
+                    epk_sq[0] = epk_sq[0] * (dx[0] ** 2)
+                    # calc intens and sig
+                    th = pk.getScattering() / 2
+                    wl = pk.getWavelength()
+                    L = (np.sin(th) ** 2) / (wl ** 4)  # at updated peak pos if applicable
+                    intens = L * np.sum(ypk[ixlo_opt:ixhi_opt])
+                    sig = L * np.sqrt(np.sum(epk_sq[ixlo_opt:ixhi_opt]))
+                    integrated = True
+            # update peak with new intens and sig
+            pk.setIntensity(intens)
+            pk.setSigmaIntensity(sig)
+            if plot_filename:
+                if not integrated:
+                    ypk = signal[peak_mask].sum(axis=0)
+                    epk_sq = np.sum(error[peak_mask] ** 2, axis=0)
+                    ixlo_opt = ixlo
+                    ixhi_opt = ixhi
+                # limits for 1D plot
+                ipad = int((ixhi - ixlo) / 2)  # extra portion of data shown outside the 1D window
+                istart = max(min(ixlo, ixlo_opt) - ipad, 0)
+                iend = min(max(ixhi, ixhi_opt) + ipad, len(xpk) - 1)
+                # 2D plot - data integrated over optimal TOF range (not range for which mask determined)
+                image_data = signal[:, :, ixlo_opt:ixhi_opt].sum(axis=2)
+                img = ax[0].imshow(image_data, norm=LogNorm(vmax=image_data[peak_mask].mean()))
+                ax[0].plot(*np.where(peak_mask.T), 'xw')
+                ax[0].plot(icol, irow, 'or')
+                title_str = f"{ipk} ({str(pk.getIntHKL())[1:-1]}) " \
```

Comment:
Might be worth putting the DetID from the peaks workspace in here rather than (or in addition to) the peak index?

Reply:
The DetID doesn't necessarily uniquely identify a peak (e.g 002 and 004 will be in the same detector) - but there's no harm in adding it as well!

Reply:
Didn't see a change for this but your observation about it not uniquely identifying the peak means it's not crucial